bookkeeper-server/src/main/java/org/apache/bookkeeper/bookie/EntryLogger.java

```diff
@@ -499,176 +576,163 @@ void createNewLog() throws IOException {
+                            + " which means that Future is waiting for acquiring lock on EntryLoggerAllocator.this");
+                    bc = allocateNewLog();
                 }
             }
+            if (entryLogPreAllocationEnabled) {
+                /*
+                 * We should submit new callable / create new instance of future only if the previous preallocation is
+                 * null or if it is done. This is needed because previous preallocation has not completed its execution
+                 * since it is waiting for lock on EntryLoggerAllocator.this.
+                 */
+                if ((preallocation == null) || preallocation.isDone()) {
+                    preallocation = allocatorExecutor.submit(new Callable<BufferedLogChannel>() {
+                        @Override
+                        public BufferedLogChannel call() throws IOException {
+                            return allocateNewLog();
+                        }
+                    });
+                }
+            }
+            LOG.info("Created new entry logger {}.", bc.getLogId());
+            return bc;
         }
 
-        BufferedLogChannel createNewLogForCompaction() throws IOException {
-            synchronized (createCompactionLogLock) {
+        synchronized BufferedLogChannel createNewLogForCompaction() throws IOException {
                 return allocateNewLog(COMPACTING_SUFFIX);
-            }
         }
 
-        private BufferedLogChannel allocateNewLog() throws IOException {
-            return allocateNewLog(".log");
+        synchronized BufferedLogChannel allocateNewLog() throws IOException {
+            return allocateNewLog(LOG_FILE_SUFFIX);
         }
 
         /**
          * Allocate a new log file.
          */
-        private BufferedLogChannel allocateNewLog(String suffix) throws IOException {
-            List<File> list = ledgerDirsManager.getWritableLedgerDirsForNewLog();
-            Collections.shuffle(list);
-            // It would better not to overwrite existing entry log files
-            File newLogFile = null;
-            do {
+        synchronized BufferedLogChannel allocateNewLog(String suffix) throws IOException {
+            File dirForNextEntryLog;
+            List<File> list;
+
+            try {
+                list = ledgerDirsManager.getWritableLedgerDirs();
+            } catch (NoWritableLedgerDirException nwe) {
+                if (!ledgerDirsManager.hasWritableLedgerDirs()) {
+                    list = ledgerDirsManager.getWritableLedgerDirsForNewLog();
+                } else {
+                    LOG.error("All Disks are not full, but getWritableLedgerDirs threw exception ", nwe);
+                    throw nwe;
+                }
+            }
+
+            dirForNextEntryLog = entryLogManager.getDirForNextEntryLog(list);
+
+            List<File> ledgersDirs = ledgerDirsManager.getAllLedgerDirs();
+            String logFileName;
+            while (true) {
```

Comment:
the current logic with allocateNewLog is not completely correct for choosing the logId and the dir for newlog.
lets assume there are multiple ledgerdirs - dir1 and dir2, current value of preallocatedLogId is 5 and for what so ever reasons lets assume that there is a file with name "6.log" in dir2. Now, when we get call for allocateNewLog, with the current logic it would end up creating new entrylog with name "6.log" in dir1 though there is a file with same name in dir2. 

With my current code, I fixed it by making sure that there is no file under any ledgerdir with that logFileName before finalizing logFileName.

Reply:
- can you have a test case reproduce the behavior? I am not seeing how the case you described here can happen with original logic.

- if there is a problem with this logic, I think it is better to fix in a subsequent change. It is not really related to abstract EntryLogManager.